### DISCUSSION TITLE
Support for TF-A Firmware Update with FWU metadata

### DISCUSSION THREAD

**Discussion by kmaincent** [rauc/rauc#1426] [Ideas]
Is there any work in progress tackling the support for PSA Firmware Update in Rauc?

If it is not the case I would like to raise the discussion about this support.

Here are the specs and the T-A documentation about PSA FWU:
https://developer.arm.com/documentation/den0118/latest/
https://trustedfirmware-a.readthedocs.io/en/latest/components/firmware-update.html

Here is some work by Sughosh Ganu from Linaro updating FWU metadata to version 2 in TF-A (merged) and U-boot (in progress):
https://review.trustedfirmware.org/q/topic:%22topics/fwu_metadata_v2_migration%22
https://lists.denx.de/pipermail/u-boot/2024-February/545732.html
Also work for FWU multi bank support features in Uboot: 
https://www.mail-archive.com/u-boot@lists.denx.de/msg431187.html

On RAUC side I can't find anything so we want to tackle this task with @rgantois and @tpetazzoni.
I am not a RAUC expert so I might ask dumb questions ;)

The aim is to be able to update the firmware + bootloader alongside the update of the kernel + rootfs.
So adding a new bootchoser won't work we need to add a new core for fw update.
Would such RAUC configuration work for you?
```
[system]
compatible=rauc-demo-arm
bootloader=uboot
bundle-formats=-plain
firmware-update=tfa-psa-fwu

[keyring]
path=demo.cert.pem

[slot.bootloader.0]
device=/dev/sdax
type=raw
fwboot=0

[slot.bootloader.1]
device=/dev/sdax
type=raw
fwboot=1

[slot.rootfs.0]
device=/dev/sdax
type=ext4
bootname=A

[slot.rootfs.1]
device=/dev/sdax
type=ext4
bootname=B
```

We will also develop a tool to update fwu metadata like ubootenv tools.

If you have any pointers or ideas please go ahead to help us with the design.

**Comment by jluebbe** [comment]
To be honest, I've not fully understood the motivation behind the FWU standards, so please excuse any misunderstandings.

If I get it right, it basically adds fallback handling functionality to the part of TF-A running as BL2. Based on persistent metadata, it will select a bank to load the rest of the firmware from and optionally decrement counters during a trail stage. This means that BL2 itself **cannot** be updated via FWU, right? Does it also mean that there needs to be a stable ABI between BL2 and the rest of the firmware?

In RAUC, we currently handle bootloader ("firmware") updates only as [*atomic* updates](https://rauc.readthedocs.io/en/latest/advanced.html#updating-the-bootloader), so there is no fallback or trail stage. Fallback is only used for the main OS and implemented in the bootloader (e.g. U-Boot, Barebox, GRUB or UEFI). By introducing fallback for the firmware as well, there are now several update sequences possible. I'm not sure which ones are sensible in practice.

1. The FW and OS can be updated together from as single bundle. Now, if FW or OS triggers a fallback, what should be done about the other? 
2. The new OS version could include the FW update as part of the rootfs and trigger it (perhaps without involving RAUC) on first boot after confirming the OS update. In this approach, the new FW would not need to support booting the old OS.
3. Some thing else?

To me, variant 2 seems simpler than 1, as there are fewer version combinations to test. Which update sequences to you want to support?

Which software component does actually perform the write to persistent storage in the FWU world? If this can be done from userspace, it might be useful to model the banks as separate slots in RAUC.

**Reply by kmaincent** [reply]
Hello,

Thanks for your answer.
You are right BL2 will boot the right bootloader according to the FWU metadata configurations.
It will also manage a fallback mechanism in case of a faulty update.
Not sure that it needs stable ABI between BL2 and the rest of the firmwares because these are different parts of the boot process. Maybe I am wrong as I still haven't investigated the FWU standards.

The aim is to be able to update FW or OS separately with RAUC. Or both together if needed.
So we would have two update mechanisms that can run during a full update.
As described in my example `firmware-update=tfa-psa-fwu` (similar to `bootloader=uboot`) will tell which FW update mechanism to use. We will use firmware slots to describe the FW location.
What do you think?

There is already a software component existing but it is written in Python :/
https://github.com/CGUSTM/fwu_gen_metadata
We should rewrite it in C and have it behave like libubootenv.

**Comment by tpetazzoni** [comment]
Hello,

After many months, we would like to revive this discussion. I had a quick chat about it with @jluebbe during Embedded Open Source Summit Europe in Vienna, but I didn't follow up right after that.

So here are a few answers to the previous questions from @jluebbe:

- You are correct that the FWU standard/mechanism is not intended to allow upgrading BL2. It is intended to allow updating what comes *after* BL2, i.e typically the FIP image that bundles BL32, BL33, etc, and the FWU metadata tells BL2 which FIP image to use. If BL2 needs to be updated, it's a separate matter which needs to take into account the specificities of the SoC in terms of how it decides whether a boot image is correct or not and how it falls back.

- The FWU mechanism indeed assumes that the interface/ABI between BL2 and what comes afterwards is stable. If you take a look at https://developer.arm.com/documentation/den0118/latest/, and specifically the diagram section 2.1.1, it shows a "BMC" bloock (in our case BL2), which is assumed to be immutable. So it looks like the idea is that firmware who want to use/implement FWU should ensure that the interface/ABI between the immutable part and the updatable-part is stable. Of course, one can think this is a dream, and can separately plan for a SoC-specific mechanism to also update BL2, but that's a separate matter.

- We definitely would want to be able to update the OS (kernel/rootfs) and the FW independently. Updates to the FW are expected to be a lot less frequent than OS updates, so it does make sense to update them separately, just like very often the OS and the "user application" are stored separately, so that they can be updated separately, with the assumption that the "user application" will need to be updated much more frequently than the OS.

- Regarding the ABI concern between the FW and the OS (new OS is flashed, but the FW is too old), the assumption is that the system developer should be careful enough when deploying updates, and if a new OS is included in an update, then that update should also include an update to the FW.

- Your proposed approach number 2 looks simpler, but doing the FW update outside of RAUC doesn't look very nice. Our whole idea in opening this discussion was precisely to see how we could make this FW update logic fit in RAUC, rather than having yet another third party thing just to handle this FW update problem. It seems to make so much more sense to have this integrated in RAUC, by modeling the banks as separate slots in RAUC like you suggest.

- If I remember correctly, at Embedded Open Source Summit Europe, you had pointed out to me that you did not understand why FWU was needed in the first place. Your assumption (again if I recall well) was that BL2 has two copies, FIP has two copies, and having FWU metadata to decide which FIP copy to use was useless, as BL2 could implement a simple scheme such as "if BL2 copy 1 was started, start FIP copy 1", if "BL2 copy 2 was started, start FIP copy 2". However, this prevents having a good recovery/fallback mechanism. In the case of the STM32MP1 platform which is our primary target for this work, it is not 100% clear how the ROM code decides whether to use copy 1 or copy 2 of BL2, but it looks like it's just about "I'm able to load it". However, the BL2 itself is able to have a much more thorough recovery mechanism, as it can arm a watchdog, start BL32/BL33, and if the watchdog is not kicked in properly, it can decide using FWU metadata to use the other copy of the FIP image, even if we're still using the same BL2. So from our perspective, enforcing "copy X of BL2 booted => copy X of FIP image" is not a good idea.

What do you think? What are the next steps to move forward with this topic? We have set it aside for several months, but we would like to actively resume the work.
